pam-1.1.1/modules/pam_access/pam_access.8
.\"     Title: pam_access
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: 06/16/2009
.\"    Manual: Linux-PAM Manual
.\"    Source: Linux-PAM Manual
.\"  Language: English
.\"
.TH "PAM_ACCESS" "8" "06/16/2009" "Linux-PAM Manual" "Linux-PAM Manual"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
pam_access \- PAM module for logdaemon style login access control
.SH "Synopsis"
.fam C
.HP \w'\fBpam_access\&.so\fR\ 'u
\fBpam_access\&.so\fR [debug] [nodefgroup] [noaudit] [accessfile=\fIfile\fR] [fieldsep=\fIsep\fR] [listsep=\fIsep\fR]
.fam
.SH "DESCRIPTION"
.PP
The pam_access PAM module is mainly for access management\&. It provides logdaemon style login access control based on login names, host or domain names, internet addresses or network numbers, or on terminal line names in case of non\-networked logins\&.
.PP
By default rules for access management are taken from config file
\FC/etc/security/access\&.conf\F[]
if you don\'t specify another file\&.
.PP
If Linux PAM is compiled with audit support the module will report when it denies access based on origin (host or tty)\&.
.SH "OPTIONS"
.PP
\fBaccessfile=\fR\fB\fI/path/to/access\&.conf\fR\fR
.RS 4
Indicate an alternative
\FCaccess\&.conf\F[]
style configuration file to override the default\&. This can be useful when different services need different access lists\&.
.RE
.PP
\fBdebug\fR
.RS 4
A lot of debug information is printed with
\fBsyslog\fR(3)\&.
.RE
.PP
\fBnoaudit\fR
.RS 4
Do not report logins from disallowed hosts and ttys to the audit subsystem\&.
.RE
.PP
\fBfieldsep=\fR\fB\fIseparators\fR\fR
.RS 4
This option modifies the field separator character that pam_access will recognize when parsing the access configuration file\&. For example:
\fBfieldsep=|\fR
will cause the default `:\' character to be treated as part of a field value and `|\' becomes the field separator\&. Doing this may be useful in conjunction with a system that wants to use pam_access with X based applications, since the
\fBPAM_TTY\fR
item is likely to be of the form "hostname:0" which includes a `:\' character in its value\&. But you should not need this\&.
.RE
.PP
\fBlistsep=\fR\fB\fIseparators\fR\fR
.RS 4
This option modifies the list separator character that pam_access will recognize when parsing the access configuration file\&. For example:
\fBlistsep=,\fR
will cause the default ` \' (space) and `\et\' (tab) characters to be treated as part of a list element value and `,\' becomes the only list element separator\&. Doing this may be useful on a system with group information obtained from a Windows domain, where the default built\-in groups "Domain Users", "Domain Admins" contain a space\&.
.RE
.PP
\fBnodefgroup\fR
.RS 4
User tokens which are not enclosed in parentheses will not be matched against the group database\&. The backwards compatible default is to try the group database match even for tokens not enclosed in parentheses\&.
.RE
.SH "MODULE TYPES PROVIDED"
.PP
All module types (\fBauth\fR,
\fBaccount\fR,
\fBpassword\fR
and
\fBsession\fR) are provided\&.
.SH "RETURN VALUES"
.PP
PAM_SUCCESS
.RS 4
Access was granted\&.
.RE
.PP
PAM_PERM_DENIED
.RS 4
Access was not granted\&.
.RE
.PP
PAM_IGNORE
.RS 4

\fBpam_setcred\fR
was called which does nothing\&.
.RE
.PP
PAM_ABORT
.RS 4
Not all relevant data or options could be gotten\&.
.RE
.PP
PAM_USER_UNKNOWN
.RS 4
The user is not known to the system\&.
.RE
.SH "FILES"
.PP
\FC/etc/security/access\&.conf\F[]
.RS 4
Default configuration file
.RE
.SH "SEE ALSO"
.PP

\fBaccess.conf\fR(5),
\fBpam.d\fR(5),
\fBpam\fR(8)\&.
.SH "AUTHORS"
.PP
The logdaemon style login access control scheme was designed and implemented by Wietse Venema\&. The pam_access PAM module was developed by Alexei Nogin <alexei@nogin\&.dnttm\&.ru>\&. The IPv6 support and the network(address) / netmask feature was developed and provided by Mike Becher <mike\&.becher@lrz\-muenchen\&.de>\&.


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



pam-0.76/Linux-PAM/modules/pam_access/pam_access.c
/* pam_access module */

/*
 * Written by Alexei Nogin <alexei@nogin.dnttm.ru> 1997/06/15
 * (I took login_access from logdaemon-5.6 and converted it to PAM
 * using parts of pam_time code.)
 *
 ************************************************************************ 
 * Copyright message from logdaemon-5.6 (original file name DISCLAIMER)
 ************************************************************************ 
 * Copyright 1995 by Wietse Venema. All rights reserved. Individual files 
 * may be covered by other copyrights (as noted in the file itself.) 
 * 
 * This material was originally written and compiled by Wietse Venema at 
 * Eindhoven University of Technology, The Netherlands, in 1990, 1991, 
 * 1992, 1993, 1994 and 1995. 
 * 
 * Redistribution and use in source and binary forms are permitted 
 * provided that this entire copyright notice is duplicated in all such 
 * copies. 
 * 
 * This software is provided "as is" and without any expressed or implied 
 * warranties, including, without limitation, the implied warranties of 
 * merchantibility and fitness for any particular purpose. 
 *************************************************************************
 */

#include <security/_pam_aconf.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <stdarg.h>
#include <syslog.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include <grp.h>
#include <errno.h>
#include <ctype.h>
#include <sys/utsname.h>

#ifndef BROKEN_NETWORK_MATCH
# include <netdb.h>
# include <sys/socket.h>
#endif

/*
 * here, we make definitions for the externally accessible functions
 * in this file (these definitions are required for static modules
 * but strongly encouraged generally) they are used to instruct the
 * modules include file to define their prototypes.
 */

#define PAM_SM_ACCOUNT

#include <security/_pam_macros.h>
#include <security/pam_modules.h>

int strcasecmp(const char *s1, const char *s2);

/* login_access.c from logdaemon-5.6 with several changes by A.Nogin: */

 /*
  * This module implements a simple but effective form of login access
  * control based on login names and on host (or domain) names, internet
  * addresses (or network numbers), or on terminal line names in case of
  * non-networked logins. Diagnostics are reported through syslog(3).
  * 
  * Author: Wietse Venema, Eindhoven University of Technology, The Netherlands.
  */

#if !defined(MAXHOSTNAMELEN) || (MAXHOSTNAMELEN < 64)
#undef MAXHOSTNAMELEN
#define MAXHOSTNAMELEN 256
#endif

#ifdef DEFAULT_CONF_FILE
# define PAM_ACCESS_CONFIG DEFAULT_CONF_FILE
#else
# define PAM_ACCESS_CONFIG "/etc/security/access.conf"
#endif

 /* Delimiters for fields and for lists of users, ttys or hosts. */

static const char *fs = ":";			/* field separator */
static const char sep[] = ", \t";		/* list-element separator */

 /* Constants to be used in assignments only, not in comparisons... */

#define YES             1
#define NO              0

 /*
  * A structure to bundle up all login-related information to keep the
  * functional interfaces as generic as possible.
  */
struct login_info {
    struct passwd *user;
    char   *from;
    const char *config_file;
    const char *service;
};

/* --- static functions for checking whether the user should be let in --- */

static void _log_err(const char *format, ... )
{
    va_list args;

    va_start(args, format);
    openlog("pam_access", LOG_CONS|LOG_PID, LOG_AUTH);
    vsyslog(LOG_ERR, format, args);
    va_end(args);
    closelog();
}

/* Parse module config arguments */

static int parse_args(struct login_info *loginfo, int argc, const char **argv)
{
    int i;

    for (i=0; i<argc; ++i) {
	if (!strncmp("fieldsep=", argv[i], 9)) {

	    /* the admin wants to override the default field separators */
	    fs = argv[i]+9;

	} else if (!strncmp("accessfile=", argv[i], 11)) {
	    FILE *fp = fopen(11 + argv[i], "r");

	    if (fp) {
		loginfo->config_file = 11 + argv[i];
		fclose(fp);
	    } else {
		_log_err("for service [%s] failed to open accessfile=[%s]"
			 , loginfo->service, 11 + argv[i]);
		return 0;
	    }
	    
	} else {
	    _log_err("unrecognized option [%s]", argv[i]);
	}
    }
    
    return 1;  /* OK */
}

typedef int match_func (char *, struct login_info *); 

static int list_match (char *, struct login_info *,
		             match_func *);
static int user_match (char *, struct login_info *);
static int from_match (char *, struct login_info *);
static int string_match (char *, char *);

/* login_access - match username/group and host/tty with access control file */

static int login_access(struct login_info *item)
{
    FILE   *fp;
    char    line[BUFSIZ];
    char   *perm;		/* becomes permission field */
    char   *users;		/* becomes list of login names */
    char   *froms;		/* becomes list of terminals or hosts */
    int     match = NO;
    int     end;
    int     lineno = 0;		/* for diagnostics */

    /*
     * Process the table one line at a time and stop at the first match.
     * Blank lines and lines that begin with a '#' character are ignored.
     * Non-comment lines are broken at the ':' character. All fields are
     * mandatory. The first field should be a "+" or "-" character. A
     * non-existing table means no access control.
     */

    if ((fp = fopen(item->config_file, "r"))!=NULL) {
	while (!match && fgets(line, sizeof(line), fp)) {
	    lineno++;
	    if (line[end = strlen(line) - 1] != '\n') {
		_log_err("%s: line %d: missing newline or line too long",
		       item->config_file, lineno);
		continue;
	    }
	    if (line[0] == '#')
		continue;			/* comment line */
	    while (end > 0 && isspace(line[end - 1]))
		end--;
	    line[end] = 0;			/* strip trailing whitespace */
	    if (line[0] == 0)			/* skip blank lines */
		continue;
	    /* Allow trailing : in last field for froms */
	    	    if (!(perm = strtok(line, fs))
		|| !(users = strtok((char *) 0, fs))
		|| !(froms = strtok((char *) 0, fs))) {
		_log_err("%s: line %d: bad field count",
			 item->config_file, lineno);
		continue;
	    }
	    if (perm[0] != '+' && perm[0] != '-') {
		_log_err("%s: line %d: bad first field",
			 item->config_file, lineno);
		continue;
	    }
	    match = (list_match(froms, item, from_match)
		     && list_match(users, item, user_match));
	}
	(void) fclose(fp);
    } else if (errno != ENOENT) {
	_log_err("cannot open %s: %m", item->config_file);
    }
    return (match == 0 || (line[0] == '+'));
}

/* list_match - match an item against a list of tokens with exceptions */

static int list_match(char *list, struct login_info *item, match_func *match_fn)
{
    char   *tok;
    int     match = NO;

    /*
     * Process tokens one at a time. We have exhausted all possible matches
     * when we reach an "EXCEPT" token or the end of the list. If we do find
     * a match, look for an "EXCEPT" list and recurse to determine whether
     * the match is affected by any exceptions.
     */

    for (tok = strtok(list, sep); tok != 0; tok = strtok((char *) 0, sep)) {
	if (strcasecmp(tok, "EXCEPT") == 0)	/* EXCEPT: give up */
	    break;
	if ((match = (*match_fn) (tok, item)))	/* YES */
	    break;
    }
    /* Process exceptions to matches. */

    if (match != NO) {
	while ((tok = strtok((char *) 0, sep)) && strcasecmp(tok, "EXCEPT"))
	     /* VOID */ ;
	if (tok == 0 || list_match((char *) 0, item, match_fn) == NO)
	    return (match);
    }
    return (NO);
}

/* myhostname - figure out local machine name */

static char * myhostname(void)
{
    static char name[MAXHOSTNAMELEN + 1];

    gethostname(name, MAXHOSTNAMELEN);
    name[MAXHOSTNAMELEN] = 0;
    return (name);
}

/* netgroup_match - match group against machine or user */

static int netgroup_match(char *group, char *machine, char *user)
{
    static char *mydomain = 0;

    if (mydomain == 0)
	yp_get_default_domain(&mydomain);
    return (innetgr(group, machine, user, mydomain));
}

/* user_match - match a username against one token */

static int user_match(char *tok, struct login_info *item)
{
    char   *string = item->user->pw_name;
    struct login_info fake_item;
    struct group *group;
    int     i;
    char   *at;

    /*
     * If a token has the magic value "ALL" the match always succeeds.
     * Otherwise, return YES if the token fully matches the username, if the
     * token is a group that contains the username, or if the token is the
     * name of the user's primary group.
     */

    if ((at = strchr(tok + 1, '@')) != 0) {	/* split user@host pattern */
	*at = 0;
	fake_item.from = myhostname();
	return (user_match(tok, item) && from_match(at + 1, &fake_item));
    } else if (tok[0] == '@') {			/* netgroup */
	return (netgroup_match(tok + 1, (char *) 0, string));
    } else if (string_match(tok, string)) {	/* ALL or exact match */
	return (YES);
    } else if ((group = getgrnam(tok))) {	/* try group membership */
	if (item->user->pw_gid == group->gr_gid)
	    return (YES);
	for (i = 0; group->gr_mem[i]; i++)
	    if (strcasecmp(string, group->gr_mem[i]) == 0)
		return (YES);
    }
    return (NO);
}

/* from_match - match a host or tty against a list of tokens */

static int from_match(char *tok, struct login_info *item)
{
    char   *string = item->from;
    int     tok_len;
    int     str_len;

    /*
     * If a token has the magic value "ALL" the match always succeeds. Return
     * YES if the token fully matches the string. If the token is a domain
     * name, return YES if it matches the last fields of the string. If the
     * token has the magic value "LOCAL", return YES if the string does not
     * contain a "." character. If the token is a network number, return YES
     * if it matches the head of the string.
     */

    if (tok[0] == '@') {			/* netgroup */
	return (netgroup_match(tok + 1, string, (char *) 0));
    } else if (string_match(tok, string)) {	/* ALL or exact match */
	return (YES);
    } else if (tok[0] == '.') {			/* domain: match last fields */
	if ((str_len = strlen(string)) > (tok_len = strlen(tok))
	    && strcasecmp(tok, string + str_len - tok_len) == 0)
	    return (YES);
    } else if (strcasecmp(tok, "LOCAL") == 0) {	/* local: no dots */
	if (strchr(string, '.') == 0)
	    return (YES);
#ifdef BROKEN_NETWORK_MATCH
    } else if (tok[(tok_len = strlen(tok)) - 1] == '.'	/* network */
	       && strncmp(tok, string, tok_len) == 0) {
	return (YES);
#else /*  BROKEN_NETWORK_MATCH */
    } else if (tok[(tok_len = strlen(tok)) - 1] == '.') {
        /*
           The code below does a more correct check if the address specified
           by "string" starts from "tok".
                               1998/01/27  Andrey V. Savochkin <saw@msu.ru>
         */

        struct hostent *h;
        char hn[3+1+3+1+3+1+3+1+1];
        int r;

        h = gethostbyname(string);
        if (h == NULL)
	    return (NO);
        if (h->h_addrtype != AF_INET)
	    return (NO);
        if (h->h_length != 4)
	    return (NO); /* only IPv4 addresses (SAW) */
        r = snprintf(hn, sizeof(hn), "%u.%u.%u.%u.",
		     (unsigned char)h->h_addr[0], (unsigned char)h->h_addr[1],
		     (unsigned char)h->h_addr[2], (unsigned char)h->h_addr[3]);
        if (r < 0 || r >= sizeof(hn))
	    return (NO);
        if (!strncmp(tok, hn, tok_len))
	    return (YES);
#endif /*  BROKEN_NETWORK_MATCH */
    }
    return (NO);
}

/* string_match - match a string against one token */

static int string_match(char *tok, char *string)
{

    /*
     * If the token has the magic value "ALL" the match always succeeds.
     * Otherwise, return YES if the token fully matches the string.
     */

    if (strcasecmp(tok, "ALL") == 0) {		/* all: always matches */
	return (YES);
    } else if (strcasecmp(tok, string) == 0) {	/* try exact match */
	return (YES);
    }
    return (NO);
}

/* end of login_access.c */

int strcasecmp(const char *s1, const char *s2) 
{
    while ((toupper(*s1)==toupper(*s2)) && (*s1) && (*s2)) {s1++; s2++;}
    return(toupper(*s1)-toupper(*s2));
}

/* --- public account management functions --- */

PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh,int flags,int argc
		     ,const char **argv)
{
    struct login_info loginfo;
    const char *user=NULL, *service=NULL;
    char *from=NULL;
    struct passwd *user_pw;

    if ((pam_get_item(pamh, PAM_SERVICE, (const void **)&service)
	!= PAM_SUCCESS) || (service == NULL) || (*service == ' ')) {
	_log_err("cannot find the service name");
	return PAM_ABORT;
    }

    /* set username */

    if (pam_get_user(pamh, &user, NULL) != PAM_SUCCESS || user == NULL
	|| *user == '\0') {
	_log_err("cannot determine the user's name");
	return PAM_USER_UNKNOWN;
    }

    /* remote host name */

    if (pam_get_item(pamh, PAM_RHOST, (const void **)&from)
	!= PAM_SUCCESS) {
	_log_err("cannot find the remote host name");
	return PAM_ABORT;
    }

    if ((from==NULL) || (*from=='\0')) {

        /* local login, set tty name */

        if (pam_get_item(pamh, PAM_TTY, (const void **)&from) != PAM_SUCCESS
            || from == NULL) {
            D(("PAM_TTY not set, probing stdin"));
	    from = ttyname(STDIN_FILENO);
	    if (from == NULL) {
	        _log_err("couldn't get the tty name");
	        return PAM_ABORT;
	     }
	    if (pam_set_item(pamh, PAM_TTY, from) != PAM_SUCCESS) {
	        _log_err("couldn't set tty name");
	        return PAM_ABORT;
	     }
        }
	if (from[0] == '/') { /* full path */
	    from++;
	    from = strchr(from, '/');
	    from++;
	}
    }

    if ((user_pw=getpwnam(user))==NULL) return (PAM_USER_UNKNOWN);

    /*
     * Bundle up the arguments to avoid unnecessary clumsiness later on.
     */
    loginfo.user = user_pw;
    loginfo.from = from;
    loginfo.service = service;
    loginfo.config_file = PAM_ACCESS_CONFIG;

    /* parse the argument list */

    if (!parse_args(&loginfo, argc, argv)) {
	_log_err("failed to parse the module arguments");
	return PAM_ABORT;
    }

    if (login_access(&loginfo)) {
	return (PAM_SUCCESS);
    } else {
	_log_err("access denied for user `%s' from `%s'",user,from);
	return (PAM_PERM_DENIED);
    }
}

/* end of module definition */

#ifdef PAM_STATIC

/* static module data */

struct pam_module _pam_access_modstruct = {
    "pam_access",
    NULL,
    NULL,
    pam_sm_acct_mgmt,
    NULL,
    NULL,
    NULL
};
#endif

